### ISSUE TITLE
Confusing error for invalid feature names

### ISSUE THREAD

**Issue by Boddlnagg** [rust-lang/cargo#3069]
I just tried to publish a crate containing features with names that have dots in them (see e.g. [here](https://github.com/contextfree/winrt-rust/blob/3ec231209a224edc92e913cf3fab1a3aa0e7d454/Cargo.toml)).
On `cargo publish` I got the following: `api errors: invalid upload request: ApplicationError("invalid crate name specified: windows.web")`, and there are actually several issues:

1. The message confusingly complains about the *crate* name
2. It should complain earlier, not on `publish`, when you think that everything should work
3. Why are dots not allowed?

By the way this also affects @retep998's 0.3 rewrite of winapi, where I originally got the idea for these features names: https://github.com/retep998/winapi-rs/blob/dev/Cargo.toml

**Comment by retep998** [comment]
This issue should probably be in https://github.com/rust-lang/crates.io

**Comment by Boddlnagg** [comment]
I think the right people will see it here, and at least my point (2) is directly related to Cargo itself.

**Comment by retep998** [comment]
cc @alexcrichton

**Comment by retep998** [comment]
It appears that the list of features in a crate on crates.io is actually a `HashMap<CrateName, ...>` which means the feature name is being interpreted as a crate name, hence why the error says it is an invalid crate name. https://github.com/rust-lang/crates.io/blob/bc4f1d79dcaac2b44e51fb7f67751d2bacca0453/src/upload.rs#L16

**Comment by alexcrichton** [comment]
Yes the actual change that needs to be made here is on crates.io, which is where this error message originates from.

**Comment by Boddlnagg** [comment]
Do I understand this correctly that there is no hope that more characters will be allowed in feature names at some point?

**Comment by alexcrichton** [comment]
@Boddlnagg currently we have no plans to allow more characters, but that doesn't necessarily mean we never will!

**Comment by dwijnand** [comment]
@Boddlnagg Given the fix needs to happen in crates.io, is there a ticket tracking this in crates.io?

**Comment by epage** [comment]
Based on this discussion, I'm moving this over to crates.io.

At this point, cargo can't directly add new validation to feature names.  We could offer a transition on an edition but then we'd need a way to refer to the old feature names, much like Rust did with keywords.
